wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic.hs

```diff
@@ -592,28 +582,18 @@ nLargestFromListBy f n = \xs ->
 
 -- | Proportionally divide the fee over each output.
 --
--- There's a special 'edge-case' when the given input is a singleton list
--- with one 0 coin. This is artifically created during input selection when
--- the transaction's amount matches exactly the source's balance.
--- In such case, we can't really compute any ratio for fees and simply return
--- the whole fee back with the given change value.
+-- Pre-condition 1: The given outputs list shouldn't be empty
+-- Pre-condition 2: None of the outputs should be null
 divvyFee :: forall dom a. CoinSelDom dom
           => (a -> Value dom) -> Fee dom -> [a] -> [(Fee dom, a)]
-divvyFee _ _   []                     = error "divvyFee: empty list"
-divvyFee f fee [a] | f a == valueZero = [(fee, a)]
-divvyFee f fee as                     = map (\a -> (feeForOut a, a)) as
+divvyFee _ _   []                             = error "divvyFee: empty list"
+divvyFee f _ as | any ((== valueZero) . f) as = error "divvyFee: some outputs are null"
```

Comment:
Do we still need this special guard? I suspect that now that we pushed the main logic fix inside `reduceChangeOutputs` , we can try to be more liberal here?

Reply:
Well, this is an invariant we want to enforce. If we don't have that, then we might get a division by zero silently coerced to `Word64` and turned to `0`. 

```hs
λ> ceiling (0 * (14 / 0 :: Double)) :: Word64
```

So, I'll better leave a guard that explicitly fails and tell us that we did something wrong :+1: (like it seems to happen in the unit test :sweat_smile: )